com.unity.perception/Runtime/GroundTruth/Labelers/KeypointLabeler.cs

```diff
@@ -108,14 +107,62 @@ bool AreEqual(Color32 lhs, Color32 rhs)
             return lhs.r == rhs.r && lhs.g == rhs.g && lhs.b == rhs.b && lhs.a == rhs.a;
         }
 
+        bool PixelOnScreen(int x, int y, (int x, int y) dimensions)
+        {
+            return x >= 0 && x < dimensions.x && y >= 0 && y < dimensions.y;
+        }
+
+        bool PixelsMatch(int x, int y, Color32 idColor, (int x, int y) dimensions, NativeArray<Color32> data)
+        {
+            var h = dimensions.y - y;
+            var pixelColor = data[h * dimensions.x + x];
+            return AreEqual(pixelColor, idColor);
+        }
+
+        static int s_PixelTolerance = 2;
+
+        // Determine the state of a keypoint. A keypoint is considered visible (state = 2) if it is on screen and not occluded
+        // by another object. The way that we determine if a point is occluded is by checking the pixel location of the keypoint
+        // against the instance segmentation mask for the frame. The instance segmentation mask provides the instance id of the
+        // visible object at a pixel location. Which means, if the keypoint does not match the visible pixel, then another
+        // object is in front of the keypoint occluding it from view. An important note here is that the keypoint has no mass, which
+        // can lead to false negatives due to rounding issues if the keypoint is on the edge of an object. Because of this we
+        // will test not only the keypoint pixel, but also the immediate surrounding pixels  to determine if the pixel
+        // is really visible. This method returns 1 if the pixel is not visible but on screen, and 0 if the pixel is off of the screen.
+        int DetermineKeypointState(Keypoint keypoint, Color32 instanceIdColor, (int x, int y) dimensions, NativeArray<Color32> data)
+        {
+            if (keypoint.state == 0) return 0;
+
+            var centerX = (int)keypoint.x;
+            var centerY = (int)keypoint.y;
+
+            var pixelOnScreen = false;
+
+            for (var x = centerX - s_PixelTolerance; x <= centerX + s_PixelTolerance; x++)
+            {
+                for (var y = centerY - s_PixelTolerance; y <= centerY + s_PixelTolerance; y++)
+                {
+                    if (!PixelOnScreen(x, y, dimensions)) continue;
+
+                    pixelOnScreen = true;
```

Comment:
This is an awkward thing to document - a keypoint is on the screen if it is outside the image by 2 pixels?